persistent/Database/Persist/Quasi/Internal/ModelParser.hs

```diff
@@ -0,0 +1,519 @@
+                , parentheticalInner
+                , some $ contentChar <|> char '(' <|> char ')'
+                ]
+        pure $ Equality (Text.pack lhs) (Text.pack rhs)
+  where
+    parentheticalInner = do
+        str <- parenthetical'
+        pure . init . drop 1 $ str
+
+sqlTypeName :: Parser String
+sqlTypeName =
+    some $
+        choice
+            [ alphaNumChar
+            , char '_'
+            ]
+
+sqlLiteral :: Parser String
+sqlLiteral = label "SQL literal" $ do
+    quote <- L.lexeme spaceConsumer $ char '\'' *> manyTill charLiteral (char '\'')
+    st <- optional $ do
+        colons <- string "::"
+        tn <- sqlTypeName
+        pure $ colons <> tn
+    pure $
+        mconcat
+            [ "'"
+            , quote
+            , "'"
+            , fromMaybe "" st
+            ]
+
+quotation :: Parser Token
+quotation = label "quotation" $ do
+    str <- L.lexeme spaceConsumer quotation'
+    pure . Quotation $ Text.pack str
+
+quotation' :: Parser String
+quotation' = char '"' *> manyTill charLiteral (char '"')
+
+parenthetical :: Parser Token
+parenthetical = label "parenthetical" $ do
+    str <- L.lexeme spaceConsumer parenthetical'
+    pure . Parenthetical . Text.pack . init . drop 1 $ str
+
+parenthetical' :: Parser String
+parenthetical' = do
+    str <- between (char '(') (char ')') q
+    pure $ "(" ++ str ++ ")"
+  where
+    q = mconcat <$> some (c <|> parenthetical')
+    c = (: []) <$> choice [contentChar, nonLineSpaceChar, char '"']
+
+blockKey :: Parser Token
+blockKey = label "block key" $ do
+    fl <- upperChar
+    rl <- many alphaNumChar
+    pure . BlockKey . Text.pack $ fl : rl
+
+ptext :: Parser Token
+ptext = label "plain token" $ do
+    str <- L.lexeme spaceConsumer $ some contentChar
+    pure . PText . Text.pack $ str
+
+docComment :: Parser Token
+docComment = hidden $ do
+    _ <- string "-- |" <* hspace
+    str <- many (contentChar <|> nonLineSpaceChar)
+    pure . DocComment . Text.pack $ str
+
+comment :: Parser Token
+comment = hidden $ do
+    _ <- (string "--" <|> string "#") <* hspace
+    str <- many (contentChar <|> nonLineSpaceChar)
+    pure . Comment . Text.pack $ str
+
+indentLevelBacktrack :: Parser () -> Parser Pos
+indentLevelBacktrack sc = lookAhead $ do
+    _ <- sc
+    L.indentLevel
+
+indented :: Parser () -> Pos -> (Parser a -> Parser b) -> Parser a -> Parser b
+indented sc ref f p =
+    f . try $ do
+        ind <- indentLevelBacktrack sc
+        if ref == ind
+            then sc *> p
+            else mzero
+
+indented' :: Parser () -> Pos -> Parser a -> Parser a
+indented' sc ref = indented sc ref id
+
+flatBlock :: Parser () -> (Parser a -> Parser b) -> Parser a -> Parser b
+flatBlock sc f p = do
+    ref <- indentLevelBacktrack sc
+    indented sc ref f p
+
+docCommentBlock :: Parser (Maybe DocCommentBlock)
+docCommentBlock = do
+    lines <-
+        flatBlock indentationConsumer many $ choice [docCommentLine, commentLine]
```

Comment:
ok, hm

prior advice was to avoid trivial recomposition but you have `f` in `flatBlock` also, and then also in `docCommentBlock` - so to rewrite without `f` in `indented`, you'd need, hm,

```suggestion
flatBlock :: Parser () -> Parser a -> Parser a
flatBlock sc p = do
    ref <- indentLevelBacktrack sc
    indented sc ref p

docCommentBlock :: Parser (Maybe DocCommentBlock)
docCommentBlock = do
    lines <-
        flatBlock indentationConsumer $ many $ choice [docCommentLine, commentLine]
```